README.md
# githooked

A GitHub webhooks worker!

[![Build Status](https://api.travis-ci.org/ScottONeal/githooked.svg)](http://travis-ci.org/ScottONeal/githooked)

**githooked** is a tiny library and companion CLI tool for handling [GitHub webhooks hooks](https://help.github.com/articles/about-webhooks/). This repo is a fork of https://github.com/coreh/hookshot and was created since the author of hookshot was innactive.

When setting up webhooks in your githooked dashboard, the types of events that should be sent to githooked should be set to: 'Just the push event'

## Examples

### Library

```javascript
var githooked = require('githooked');
githooked('refs/heads/master', 'git pull && make').listen(3000)
```

### CLI Tool

```bash
githooked -r refs/heads/master 'git pull && make'
```

## Usage

The library exposes a single function, `githooked()`. When called, this functions returns an express instance configured to handle webhooks pushes from GitHub. You can react to pushes to specific branches by listening to specific events on the returned instance, or by providing optional arguments to the `githooked()` function.

```javascript
githooked()
  .on('refs/heads/master', 'git pull && make')
  .listen(3000)
```

```javascript
githooked('refs/heads/master', 'git pull && make').listen(3000)
```

## Arguments

GitHooked supports up to three arguments: reference, action, and options. The first argument is the branch reference from the GitHooked webhook (i.e: `refs/heads/master`). If only one argument is supplied, it should be the action that needs to be ran. In this instance githooked will bind to every webhook event sent from GitHub.

```js
githooked('branch/references', 'action', { /* options */});
```

### Reference

References are specific webhook references or actions fired when editing tags or branches changes happen. This argument is provided so you can bind to specific branch or the following event hooks:

 - **hook**: This will bind to all webhooks
 - **create**: fired on [create events](https://developer.github.com/v3/activity/events/types/#createevent)
 - **delete**: fired on [delete events](https://developer.github.com/v3/activity/events/types/#deleteevent)
 - **push**: fired on [push events](https://developer.github.com/v3/activity/events/types/#pushevent)

### Action

Actions can either be shell commands or JavaScript functions.

```javascript
githooked('refs/heads/master', 'git pull && make').listen(3000)
```

```javascript
githooked('refs/heads/master', function(info) {
  // do something with push info ...
}).listen(3000)
```

### Options

Lastly, the third option is an object of configuration parameters. Usage:

```js
githooked('push', 'git pull && make', {
  json: {
    limit: '100mb',
    strict: true
  },
  middleware: [
    require('connect-timeout'),
    function(req, res, next) {
      // Do something
      next();
    }
  ]
}})
```

The following configuration options are:

#### json (Object)

These are arguments passed to express's [body-parsers json() middleware](https://github.com/expressjs/body-parser#bodyparserjsonoptions)

#### middleware (Function|Array[Function])

This is an array or function of valid [express middleware](http://expressjs.com/en/guide/using-middleware.html). This middleware will be applied before any other express middleware. If an array is provided, the middleware will be applied in the order they are declared in the array.

#### secret (String)

GitHub webhooks can pass a secret which is used as an validation mechanism between your GitHub repo and your githooked server. Read more about it [here](https://developer.github.com/v3/repos/hooks/#create-a-hook). Validation of the payload will be the first operation performed on incoming requests. If using githooked for any serious purposes this option should be necessary. If validation failed an error event will be called with value of 'signature validation failed' or 'no provider signature':

```js
  githooked.on('error', function(msg) {
    if ( msg === 'signature validation failed' ) {
      // do something
    }
  })
```

#### logFile (String|stream) TODO

If your action is a shell call, then this option will log all STDOUT and STDERR into the provided stream or file descriptor

### Mounting to existing express servers

**githooked** can be mounted to a custom route on your existing express server:

```javascript
// ...
app.use('/my-github-hook', githooked('refs/heads/master', 'git pull && make'));
// ...
```

### Special Events

Special events are fired when branches/tags are created, deleted:

```javascript
githooked()
.on('create', function(info) {
  console.log('ref ' + info.ref + ' was created.')
})
.on('delete', function(info) {
  console.log('ref ' + info.ref + ' was deleted.')
})
```

The `push` event is fired when a push is made to any ref:

```javascript
githooked()
.on('push', function(info) {
  console.log('ref ' + info.ref + ' was pushed.')
})
```

Finally, the `hook` event is fired for every post-receive hook that is send by GitHub.

```javascript
githooked()
.on('push', function(info) {
  console.log('ref ' + info.ref + ' was pushed.')
})
```

### Spawn Event

If githooked was created with a shell command as the action, it will throw a spawn event with the child_process spawn instance.

```javascript
var server = githooked('refs/head/master', 'git pull && make').listen(3000);

server.on('spawn', function(spawn) {
  // Bind on close to get exit code
  spawn.on('close', function(code) {
    if ( code !== 0 ) {
      console.log('something went wrong');
    }
  });
});
```

### CLI Tool

A companion CLI tool is provided for convenience. To use it, install **githooked** via npm using the `-g` flag:

```bash
npm install -g githooked
```

The CLI tool takes as argument a command to execute upon GitHub post-receive hook:

```bash
githooked 'echo "PUSHED!"'
```

You can optionally specify an HTTP port via the `-p` flag (defaults to 3000) and a ref via the `-r` flag (defaults to all refs):

```bash
githooked -r refs/heads/master -p 9001 'echo "pushed to master!"'
```


CHANGELOG.md
1.1.2 / 2017-11-30
==================

  * Validate secret using a constant-time comparison to avoid timing attacks.

1.1.1 / 2015-02-06
==================

  * Added secret option for payload validation

1.0.4 / 2015-01-22
==================

  * Added the ability for ping events to be properly responded to

1.0.3 / 2015-01-22
==================

  * Create basic BDD test coverage
  * Added Body Parser JSON limit to be 1mb

0.1.0 / 2014-09-16
==================

  * Bump dependency versions.
  * Correctly handle both json and x-www-form-urlencoded payloads.

0.0.2 / 2013-05-16
==================

  * Implement initial Windows and Cygwin support

0.0.1 / 2013-04-22
==================

  * Initial Release


test/cli.js
'use strict';

var assert  = require('assert'),
    cp      = require('child_process'),
    path    = require('path'),
    getport = require('getport'),
    request = require('superagent'),
    cli     = path.resolve(__dirname + '/../bin/githooked');

describe('githooked cli', function() {
  it('should return help if no action is provided with exit code 1', function(done){
    cp.exec(cli + ' -p 3965', function(err) {
      assert.equal(err.code, 1);
      done();
    });
  });

  // 1.0.0 backwords compat to calling actions
  // Wow, need to find a better way to test CLIs that start a server
  it('should it should start a server and accept requests', function(done) {
    var gh;
    getport(function(err, port) {
      if ( err ) { done(err); }

      // The pain difference in the following test is how actions (-a) are passed to githooked cli
      gh = cp.spawn(cli, ['-p', port, '-r', 'test', '-a', 'echo "THIS IS A TEST"']);

      var buf = '',
          started = false;

      gh.stdout.on('data', function(chunk) {
        buf += chunk;
        if ( !started && buf.match('githooked server started') ) {
          started = true;
          webhookCaller(port, done);
        }
      });

      gh.on('error', function(chunk) {
        done(new Error(chunk));
      });
    });

    // Cleanup
    after(function() {
      gh.kill();
    });
  });

  it('should support the action being undefined commander args', function(done) {
    var gh;
    getport(function(err, port) {
      if ( err ) { done(err); }

      gh = cp.spawn(cli, ['-p', port, '-r', 'test', 'echo "THIS IS A TEST"']);

      var buf = '',
          started = false;

      gh.stdout.on('data', function(chunk) {
        buf += chunk;
        if ( !started && buf.match('githooked server started') ) {
          started = true;
          webhookCaller(port, done);
        }
      });

      gh.on('error', function(chunk) {
        done(new Error(chunk));
      });
    });

    // Cleanup
    after(function() {
      gh.kill();
    });
  });
});

function webhookCaller(port, cb) {
  request
    .post('localhost:' + port + '/')
    .set('Content-Type', 'application/json')
    .send({ ref: 'test' })
    .end(function(err, res) {
      if ( res.status === 202 ) {
        cb();
      }
    });
}


test/test.js
'use strict';

var assert    = require('assert'),
    supertest = require('supertest'),
    githooked = require('../lib'),
    crypto    = require('crypto');

process.env.NODE_ENV = 'test';

describe('githooked initialization', function() {

  var server, request;

  beforeEach(function() {
     server  = githooked('test', 'exit 1');
     request = supertest(server);
  });

  it('should have property githookedServer.ghRef', function() {
    assert.equal(server.ghRef, 'test');
  });

  it('should have property githookedServer.ghAction', function() {
    assert.equal(server.ghAction, 'exit 1');
  });

  it('should have property githookedServer.ghOptions', function() {
    var options = { json: '10mb', logFile: './goothoked.log' };
    var scopedServer = githooked('test', 'exit 1', options);
    assert.equal(scopedServer.ghOptions, options);
  });

  it('defaults action to first argument if only one argument sent', function() {
    var scopedServer = githooked('exit 1');
    assert(scopedServer);
  });

  it('should throw an error when if option argument is not an object', function() {
    assert.throws(function() {
      githooked('test', 'exit 1', []);
    }, Error);
  });

  it('should accept bodyParser.json middleware options', function(done) {
    var scopedServer = githooked('test', 'exit 1', { json: { limit: '1kb'}});
    var buf = new Buffer(1024);
    buf.fill('.');

    supertest(scopedServer)
      .post('/')
      .set('Content-Type', 'application/json')
      .send(JSON.stringify({ str: buf.toString() }))
      .expect(413, done);
  });

  it('should accept a default payload size limit of 1mb', function(done) {
    var buf = new Buffer(999990);
    buf.fill('.');

    request
      .post('/')
      .set('Content-Type', 'application/json')
      .send(JSON.stringify({ ref: 'test', str: buf.toString() }))
      .expect(202, done);
  });

  it('should add a middleware function if options.middlware = function', function(done) {
    var middlewarePassed = false;

    var scopedServer = githooked('test', 'exit 1', { middleware:
       function(req, res, next) {
         middlewarePassed = true;
         next();
       }
    });

    supertest(scopedServer)
      .post('/')
      .set('Content-Type', 'application/json')
      .send(JSON.stringify({ ref: 'test' }))
      .end(function(err) {
        assert.equal(middlewarePassed, true);
        done(err);
      });
  });

  it('should accept an array of middleware functions, if optons.middleware = [function]', function(done) {
    var middleware1Passed = false,
        middleware2Passed = false;

    var scopedServer = githooked('test', 'exit 1', { middleware: [
       function(req, res, next) {
         middleware1Passed = true;
         next();
       },
       function(req, res, next) {
         middleware2Passed = true;
         next();
       },
    ]});

    supertest(scopedServer)
      .post('/')
      .set('Content-Type', 'application/json')
      .send(JSON.stringify({ ref: 'test' }))
      .end(function(err) {
        assert.equal(middleware1Passed, true);
        assert.equal(middleware2Passed, true);
        done(err);
      });

  });
});

describe('githooked secret validation', function() {

  it('should throw an error if secret is not a string', function() {
    assert.throws(function() {
      githooked('test', 'exit 1', { secret: [] });
    }, Error);
  });

  it('should emit an error if secret not sent from provider', function(done) {
    var scopedServer = githooked('test', 'exit 1', { secret: 'test-secret' });

    supertest(scopedServer)
      .post('/')
      .set('Content-Type', 'application/json')
      .send({ ref: 'test' })
      .expect(401)
      .end(function(err) {
        if ( err ) { done(err); }
      });

      scopedServer.on('error', function(msg) {
        assert.equal(msg, 'no provider signature');
        done();
      });
  });

  it('should send a 401 if signatures do not match', function(done) {
    var secret       = 'test-secret',
        body         = { ref: 'test'},
        scopedServer = githooked('test', 'exit 1', { secret: secret });

    supertest(scopedServer)
      .post('/')
      .set('Content-Type', 'application/json')
      .set('X-Hub-Signature', 'thisisabadsignature')
      .send(body)
      .expect(401)
      .end(function(err) {
        if ( err ) { done(err); }
      });

    scopedServer.on('error', function(msg) {
      assert.equal(msg, 'signature validation failed');
      done();
    });
  });

  it('should send a 202 if signature validations match', function(done) {
    var secret       = 'test-secret',
        body         = { ref: 'test'},
        scopedServer = githooked('test', 'exit 1', { secret: secret });

    supertest(scopedServer)
      .post('/')
      .set('Content-Type', 'application/json')
      .set('X-Hub-Signature', createSignature(secret, JSON.stringify(body)))
      .send(body)
      .expect(202, done);
  });
});

describe('githooked server', function() {

  var server, request;

  beforeEach(function() {
     server  = githooked('test', 'exit 1');
     request = supertest(server);
  });

  it('should send a 200 back for a PING event', function(done) {
    request
      .post('/')
      .set('X-GitHub-Event', 'ping')
      .set('Content-Type', 'application/json')
      .send({zen: 'thisisatest', hook_id: 'thisisatest', hook: 'thisisatest'})
      .expect(200, done);
  });

  it('should throw an error on invalid payload', function(done) {
    request
      .post('/')
      .set('Content-Type', 'application/json')
      .send('asdf')
      .expect(400, done);
  });

  it('should emit an error on invalid payload', function(done) {
    server.on('error', function(msg) {
      assert.equal(msg, 'invalid ref');
      done();
    });

    request
      .post('/')
      .set('Content-Type', 'application/json')
      .send({})
      .expect(400)
      .end(function(err) {
        assert(err instanceof Error);
      });
  });

  it('should get a create event', function(done) {
    server.on('create', function(payload) {
      assert(payload);
      done();
    });

    request
      .post('/')
      .set('Content-Type', 'application/json')
      .send({ ref: 'test', created: true })
      .expect(202)
      .end(function(err) {
        assert(!err);
      });
  });

  it('should get a delete event', function(done) {
    server.on('delete', function(payload) {
      assert(payload);
      done();
    });

    request
      .post('/')
      .set('Content-Type', 'application/json')
      .send({ ref: 'test', deleted: true })
      .expect(202)
      .end(function(err) {
        assert(!err);
      });
  });

  it('should get a push event if created and deleted are falsy', function(done) {
    server.on('push', function(payload) {
      assert(payload);
      done();
    });

    request
      .post('/')
      .set('Content-Type', 'application/json')
      .send({ ref: 'test' })
      .expect(202)
      .end(function(err) {
        assert(!err);
      });
  });

  it('should get reference event with payload on hook', function(done) {
    server.on('test', function() {
      done();
    });

    request
      .post('/')
      .set('Content-Type', 'application/json')
      .send({ ref: 'test' })
      .expect(202)
      .end(function(err) {
        assert(!err);
      });

  });

  it('should run command `exit 1` when running receiving a hook for reference "test"', function(done) {

    server.on('spawn', function(instance) {
      instance.on('close', function(code) {
        assert.equal(code, 1);
        done();
      });
    });

    request
      .post('/')
      .set('Content-Type', 'application/json')
      .send({ ref: 'test' })
      .expect(202)
      .end(function(err) {
        assert(!err);
      });
  });
});

function createSignature(secret, body) {
  return 'sha1=' + crypto.createHmac('sha1', secret).update(body).digest('hex');
}


lib/index.js
'use strict';

var bodyParser = require('body-parser'),
    normalize  = require('path').normalize,
    spawn      = require('child_process').spawn,
    crypto     = require('crypto'),
    express    = require('express');

module.exports = GitHooked;

/**
 * Create a express app with github webhook parsing and actions
 *
 * @param {String} This is the git reference being listened to
 * @param {String|Function} This is the action which will be invoked when the reference is called. Can be either a function that will be executed or a string which will be called from a shell
 * @api public
 */
function GitHooked(ref, action, options) {

  // default ref to hook and action to reference if only single args sent
  // This means that all webhooks will invoke the action
  if (arguments.length === 1) {
    action = ref;
    ref    = 'hook';
  }

  // Defaults options to empty object
  options = options || {};

  // Check if options is an associative array
  if ( options.toString() !== '[object Object]' ) {
    throw new Error('GitHooked: Options argument supplied, but type is not an Object');
  }

  // create express instance
  var githooked = express();

  // Setup optional middleware
  setupMiddleware(githooked, options);

  // default bodyparser.json limit to 1mb
  options.json = options.json || { limit: '1mb' };

  githooked.use(bodyParser.urlencoded({ extended: false, limit: options.json.limit || '1mb' }));

  // Setup secret signature validation
  if ( options.secret ) {

    // Throw error if secret is not a string
    if ( typeof options.secret !== 'string' ) {
      throw new Error('GitHooked: secret provided but it is not a string');
    }
    options.json.verify = function(req, res, buf) {
      signatureValidation(githooked, options, req, res, buf);
    };
  }

  // json parsing middleware limit is increased to 1mb, otherwise large PRs/pushes will
  // fail due to maiximum content-length exceeded
  githooked.use(bodyParser.json(options.json));

  // Bind Ref, Action, and Options to Express Server
  githooked.ghRef     = ref;
  githooked.ghAction  = action;
  githooked.ghOptions = options;

  // main POST handler
  githooked.post('/', function (req, res ) {
    var payload = req.body;

    // Check if ping event, send 200 if so
    if (req.headers['x-github-event'] === 'ping') {
      res.sendStatus(200);
      return;
    }

    // Check if Payload was sent
    if (!payload) {
      githooked.emit('error', 'no payload');
      return res.sendStatus(400);
    }

    // Check if payload has a ref, this is required in order to run contexted scripts
    if (!payload.ref || typeof payload.ref !== 'string') {
      githooked.emit('error', 'invalid ref');
      return res.sendStatus(400);
    }

    // Created hook event
    if (payload.created) {
      githooked.emit('create', payload);
    }

    // Deleted hook event
    else if (payload.deleted) {
      githooked.emit('delete', payload);
    }

    // Else, default to a push hook event
     else {
      githooked.emit('push', payload);
    }

    // Always emit a global 'hook' event so people can watch all requests
    githooked.emit('hook', payload);

    // Emit a ref type specfic event
    githooked.emit(payload.ref, payload);
    res.status(202);
    res.send('Accepted\n');
  });

  if (typeof action === 'string') {
    var shell = process.env.SHELL,
        args  = ['-c', action],
        opts  = { stdio: 'inherit' };

    // Windows specfic env checks
    if (shell && isCygwin()) {
      shell = cygpath(shell);
    } else if (isWin()) {
      shell = process.env.ComSpec;
      args = ['/s', '/c', '"' + action + '"'];
      opts.windowsVerbatimArguments = true;
    }

    githooked.on(ref, function() {
      // Send emit spawn event w/ instance
      githooked.emit('spawn', spawn(shell, args, opts));
    });
  }
  else if (typeof action === 'function') {
    githooked.on(ref, action);
  }

  // Development Error middleware
  if ( process.env.NODE_ENV === 'development' ) {
    githooked.use(function(err, req, res, next) {
      console.log(err.stack);
      next(err);
    });
  }

  // Default Error middlware
  githooked.use(function(err, req, res, next) { // jshint ignore:line
    githooked.emit('error', err);
    if ( !res.headersSent ) {
      res.status(500);
      res.end(err.message);
    }
  });

  return githooked;
}

function signatureValidation(githooked, options, req, res, buf) {
  // Use crypto.timingSafeEqual when available to avoid timing attacks
  // See https://codahale.com/a-lesson-in-timing-attacks/
  var compareStrings = function(a, b) { return a === b };
  var compareSecure = function(a, b) {
    return a.length === b.length && crypto.timingSafeEqual(new Buffer(a), new Buffer(b));
  }
  var validateSecret = crypto.timingSafeEqual ? compareSecure : compareStrings;

  var providedSignature = req.headers['x-hub-signature'];

  // Throw an error if secret was provided but no X-Hub-Signature header present
  if ( !providedSignature ) {
    res.sendStatus(401);
    githooked.emit('error', 'no provider signature');
    throw new Error('no provider signature');
  }

  var ourSignature = 'sha1=' + crypto.createHmac('sha1', options.secret).update(buf.toString()).digest('hex');

  // Validate Signatures
  if ( ! validateSecret(providedSignature, ourSignature) ) {
    res.sendStatus(401);
    githooked.emit('error', 'signature validation failed');
    throw new Error('signature validation failed');
  }
}

/**
 * Setup middleware if options configured properly
 *
 * @return {Undefined}
 * @api private
 */
function setupMiddleware(githooked, options) {

  // Optional middleware pass
  if ( options.middleware ) {
    // If middleware = function, convert to array
    if ( typeof options.middleware === 'function' ) {
      options.middleware = [options.middleware];
    }

    // Add middleware
    options.middleware.forEach(function(fn) {
      githooked.use(fn);
    });
  }
}

/**
 * Returns `true` if node is currently running on Windows, `false` otherwise.
 *
 * @return {Boolean}
 * @api private
 */

function isWin () {
  return 'win32' === process.platform;
}

/**
 * Returns `true` if node is currently running from within a "cygwin" environment.
 * Returns `false` otherwise.
 *
 * @return {Boolean}
 * @api private
 */

function isCygwin () {
  // TODO: implement a more reliable check here...
  return isWin() && /cygwin/i.test(process.env.HOME);
}

/**
 * Convert a Unix-style Cygwin path (i.e. "/bin/bash") to a Windows-style path
 * (i.e. "C:\cygwin\bin\bash").
 *
 * @param {String} path
 * @return {String}
 * @api private
 */

function cygpath (path) {
  path = normalize(path);
  if (path[0] === '\\') {
    // TODO: implement better cygwin root detection...
    path = 'C:\\cygwin' + path;
  }
  return path;
}
